### PR TITLE
Add `Default` implementation for `&CStr`, `CString`, `Path`

### PR DESCRIPTION
**Pull request by tbu-**

### CODE REVIEW

**Comment by tbu-** [comment]
Split off from #32785. Some motivation: https://github.com/rust-lang/rust/pull/32785#issuecomment-210516038.

**Comment by alexcrichton** [comment]
`CString` seems fine to me personally as `CString::new` is the "clear and obvious constructor", the others seem conventionally fine but not perhaps well motivated.

**Comment by tbu-** [comment]
`""` is the default value for `&str`, I'd argue that the equivalent value should be the default value for `&CStr`.

File: src/libstd/ffi/c_str.rs [hunk 1]
```diff
@@ -326,6 +326,22 @@ impl fmt::Debug for CStr {
     }
 }
 
+#[stable(feature = "cstr_default", since = "1.9.0")]
```
Comment (GuillaumeGomez) [inline src/libstd/ffi/c_str.rs]: We just hit 1.8.0 release, so it becomes `since = 1.10.0` (if you could confirm @alexcrichton?).

**Comment by nagisa** [comment]
I’m not sure about `""` as the default path. You can’t even do anything sensible with it (at least on linux), other than join to other paths.

**Comment by tbu-** [comment]
Dunno, maybe `base.join(config_option.unwrap_or_default())`

**Comment by nagisa** [comment]
@tbu- for that purpose an equivalent of UNIX’s `.` has the same behaviour.

**Comment by tbu-** [comment]
@nagisa Not for verbatim paths on Windows.

**Comment by aturon** [comment]
I'm fine with the `CString`-related defaults, but agree that the one for `Path` is probably more problematic than it's worth.

**Comment by alexcrichton** [comment]
The libs team discussed this PR during triage today and the conclusion was that the `Default` implementations for `CStr` and `CString` seem fine, but as mentioned previously we'd prefer that `Path` were left out for now.

Could you make those changes @tbu- and I'll r+? Thanks again for the PR!

**Comment by tbu-** [comment]
@alexcrichton Changed.

**Pull request merged**